diff -c original/sun50i-h6.dtsi latest/sun50i-h6.dtsi
*** a/arch/arm64/boot/dts/allwinner/sun50i-h6.dtsi	2021-01-03 19:08:55.062835000 +0100
--- b/arch/arm64/boot/dts/allwinner/sun50i-h6.dtsi	2021-01-03 19:41:15.106356337 +0100
***************
*** 16,21 ****
--- 16,31 ----
  	#address-cells = <1>;
  	#size-cells = <1>;
  
+ 	ac200_pwm_clk: ac200_clk {
+ 		compatible = "pwm-clock";
+ 		#clock-cells = <0>;
+ 		clock-frequency = <24000000>;
+ 		pinctrl-names = "default";
+ 		pinctrl-0 = <&pwm1_pin>;
+ 		pwms = <&pwm 1 42 0>;
+ 		status = "disabled";
+ 	};
+ 
  	cpus {
  		#address-cells = <1>;
  		#size-cells = <0>;
***************
*** 286,291 ****
--- 296,305 ----
  			cpu_speed_grade: cpu-speed-grade@1c {
  				reg = <0x1c 0x4>;
  			};
+ 
+ 			ephy_calib: ephy_calib@2c {
+ 				reg = <0x2c 0x2>;
+ 			};			
  		};
  
  		watchdog: watchdog@30090a0 {
***************
*** 330,335 ****
--- 344,354 ----
  				drive-strength = <40>;
  			};
  
+ 			i2c3_pins: i2c3-pins {
+ 				pins = "PB17", "PB18";
+ 				function = "i2c3";
+ 			};
+ 
  			hdmi_pins: hdmi-pins {
  				pins = "PH8", "PH9", "PH10";
  				function = "hdmi";
***************
*** 367,372 ****
--- 386,396 ----
  				bias-pull-up;
  			};
  
+ 			pwm1_pin: pwm1-pin {
+ 				pins = "PB19";
+ 				function = "pwm1";
+ 			};
+ 
  			mmc2_pins: mmc2-pins {
  				pins = "PC1", "PC4", "PC5", "PC6",
  				       "PC7", "PC8", "PC9", "PC10",
***************
*** 606,611 ****
--- 630,660 ----
  			#size-cells = <0>;
  		};
  
+ 		i2c3: i2c@5002c00 {
+ 			compatible = "allwinner,sun6i-a31-i2c";
+ 			reg = <0x05002c00 0x400>;
+ 			interrupts = <GIC_SPI 7 IRQ_TYPE_LEVEL_HIGH>;
+ 			clocks = <&ccu CLK_BUS_I2C3>;
+ 			resets = <&ccu RST_BUS_I2C3>;
+ 			pinctrl-names = "default";
+ 			pinctrl-0 = <&i2c3_pins>;
+ 			status = "disabled";
+ 			#address-cells = <1>;
+ 			#size-cells = <0>;
+ 
+ 			ac200: mfd@10 {
+ 				compatible = "x-powers,ac200";
+ 				reg = <0x10>;
+ 				clocks = <&ac200_pwm_clk>;
+ 
+ 				ac200_ephy: phy {
+ 					compatible = "x-powers,ac200-ephy";
+ 					nvmem-cells = <&ephy_calib>;
+ 					nvmem-cell-names = "ephy_calib";
+ 				};
+ 			};
+ 		};
+ 
  		emac: ethernet@5020000 {
  			compatible = "allwinner,sun50i-h6-emac",
  				     "allwinner,sun50i-a64-emac";
